### PR TITLE
fix memory leak

### PR DESCRIPTION
**Pull request by Gorash**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/js/views/abstract_view.js [hunk 1]
```diff
@@ -133,20 +133,19 @@ var AbstractView = Class.extend({
...
             var Controller = self.Controller || self.config.Controller;
             var controllerParams = _.extend({
                 initialState: state,
             }, self.controllerParams);
-            var controller = new Controller(parent, model, renderer, controllerParams);
+            var controller = new Controller(parent, self.model, renderer, controllerParams);
             renderer.setParent(controller);
 
-            if (!self.model) {
+            if (self.model.getParent() === parent) {
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/abstract_view.js]: instead of using this heuristic, i suggest to check at line 134:
```
// check if a model already exists, as if not, one will be created and
// we'll have to set the controller as its parent
var alreadyHasModel = !!this.model;
```
and do
`if (!alreadyHasModel) {`
here.

I think that this is equivalent, but it's more explicit.

File: addons/web/static/tests/views/calendar_tests.js [hunk 2]
```diff
@@ -1599,6 +1600,64 @@ QUnit.module('Views', {
             "should be one event in the all day row");
         calendar.destroy();
     });
+
+    QUnit.test('test if the CalendarView create memory leak', function (assert) {
```
Comment (aab-odoo) [inline addons/web/static/tests/views/calendar_tests.js]: there is no memory leak in the Calendar view

File: addons/web/static/tests/views/calendar_tests.js [hunk 3]
```diff
@@ -1599,6 +1600,64 @@ QUnit.module('Views', {
...
+        assert.expect(2);
+
+        var activeWidgets = [];
+        var initWidget = mixins.ParentedMixin.init;
```
Comment (aab-odoo) [inline addons/web/static/tests/views/calendar_tests.js]: i'd name it originalInit (as this is not related to widgets)

File: addons/web/static/tests/views/calendar_tests.js [hunk 4]
```diff
@@ -1599,6 +1600,64 @@ QUnit.module('Views', {
...
+
+        calendar.destroy();
+
+        activeWidgets = _.filter(activeWidgets, function (widget) {
+            return !widget.isDestroyed();
+        });
+
+        assert.deepEqual(activeWidgets.length, 0, "every widget must be destroyed");
+
+        mixins.ParentedMixin.init = initWidget;
```
Comment (aab-odoo) [inline addons/web/static/tests/views/calendar_tests.js]: I'm not sure that this test is correct. There is (imho) what you should do:
1) create view
2) store number of living instances
3) __destroy view
4) create view
5) check that the number of living instances is the same
6) call destroy to properly destroy parent widget and restore everything
7) call unpatch

File: addons/web/static/tests/views/calendar_tests.js [hunk 5]
```diff
@@ -1599,6 +1600,64 @@ QUnit.module('Views', {
             "should be one event in the all day row");
         calendar.destroy();
     });
+
+    QUnit.test('test if the CalendarView create memory leak', function (assert) {
+        assert.expect(2);
+
+        var activeWidgets = [];
+        var initWidget = mixins.ParentedMixin.init;
+        mixins.ParentedMixin.init = function () {
```
Comment (aab-odoo) [inline addons/web/static/tests/views/calendar_tests.js]: use testUtils.patch(mixins.ParentedMixin, ...)

File: addons/web/static/tests/views/calendar_tests.js [hunk 6]
```diff
@@ -1599,6 +1600,64 @@ QUnit.module('Views', {
...
+            activeWidgets.push(this);
```
Comment (aab-odoo) [inline addons/web/static/tests/views/calendar_tests.js]: increment a counter instead (named nbInstances for example)

File: addons/web/static/tests/views/calendar_tests.js [hunk 7]
```diff
@@ -1599,6 +1600,64 @@ QUnit.module('Views', {
...
+            return initWidget.apply(this, arguments);
+        };
```
Comment (aab-odoo) [inline addons/web/static/tests/views/calendar_tests.js]: patch destroy as well, and decrement the counter (only if not destroyed yet), then call _super

File: addons/web/static/tests/views/form_tests.js [hunk 8]
```diff
@@ -6228,5 +6228,89 @@ QUnit.module('Views', {
         form.destroy();
     });
 
+    QUnit.test('test if the FormView create memory leak', function (assert) {
```
Comment (aab-odoo) [inline addons/web/static/tests/views/form_tests.js]: same remarks apply

File: addons/web/static/tests/views/kanban_tests.js [hunk 9]
```diff
@@ -1936,6 +1937,62 @@ QUnit.module('Views', {
                 "quick create should have been added in the first column");
         }
     });
+
+    QUnit.test('test if the KanbanView create memory leak', function (assert) {
+        assert.expect(2);
```
Comment (aab-odoo) [inline addons/web/static/tests/views/kanban_tests.js]: idem

File: addons/web/static/tests/views/list_tests.js [hunk 10]
```diff
@@ -3181,6 +3182,62 @@ QUnit.module('Views', {
         list.destroy();
     });
 
+    QUnit.test('test if the ListView create memory leak', function (assert) {
+        assert.expect(2);
```
Comment (aab-odoo) [inline addons/web/static/tests/views/list_tests.js]: idem

File: addons/web/static/tests/views/calendar_tests.js (hunk 6 above)
Comment (Gorash) [reply addons/web/static/tests/views/calendar_tests.js]: It's easier to fix with the widgets

**Comment by aab-odoo** [comment]
We'll rewrite the commit messages together as well

File: addons/web/static/src/js/views/abstract_view.js (hunk 1 above)
Comment (Gorash) [reply addons/web/static/src/js/views/abstract_view.js]: I'll try your solution

File: addons/web/static/tests/views/calendar_tests.js (hunk 6 above)
Comment (aab-odoo) [reply addons/web/static/tests/views/calendar_tests.js]: don't think so

File: addons/point_of_sale/static/src/js/devices.js [hunk 11]
```diff
@@ -93,8 +93,9 @@ var JobQueue = function(){
```
Comment (aab-odoo) [inline addons/point_of_sale/static/src/js/devices.js]: [FIX] web: properly set parent when using mixins

The parentedMixin doesn't automatically set the parent, so when this mixin
is used, the parent must be set manually. This wasn't done at several places,
and thus those elements weren't destroyed with their actual parent.

In master, we should probably make the ParentedMixin set the parent itself.

File: addons/web/static/src/js/views/abstract_view.js [hunk 12]
```diff
@@ -133,20 +133,19 @@ var AbstractView = Class.extend({
     getController: function (parent) {
```
Comment (aab-odoo) [inline addons/web/static/src/js/views/abstract_view.js]: [FIX] web: Views: always destroy BasicModel instances

Before this rev., it might happen that the parent of BasicModel instances
wasn't correctly set to the Controller instance. It was actually set to the
ActionManager instance, which is never destroyed. So when this happened,
the BasicModel instance wasn't destroyed when it's controller was.

File: addons/web/static/tests/views/calendar_tests.js [hunk 13]
```diff
@@ -1599,6 +1600,64 @@ QUnit.module('Views', {
...
+            },
+        };
+
+        var calendar = createView(params);
+        calendar.destroy();
+
+        activeWidgets = [];
+        calendar = createView(params);
+
+        // call read destroy method
```
Comment (aab-odoo) [inline addons/web/static/tests/views/calendar_tests.js]: call destroy function of controller to ensure that it correctly destroys everything

File: addons/web/static/tests/views/calendar_tests.js (hunk 5 above)
Comment (Gorash) [reply addons/web/static/tests/views/calendar_tests.js]: ? in saas-16

**Comment by Gorash** [comment]
@aab-odoo ok?
